### ISSUE TITLE
Maybe a code walker related issue with lagging of GUI

### ISSUE THREAD

**Issue by xpxaxsxi** [SWI-Prolog/swipl-devel#853]
I have a very simple program, why the make is taking so long, my Windows 10 shows a busy icon and GUI is not responding for a second or two. I have a lows specs laptop.

This issued has been around maybe half a year, I have been enjoying the really quick makes from Swi-Prolog for maybe 7 years.

Maybe the code walker is doing some extra work??? Maybe it is walking all possible libraries, is it caching anything? Maybe it is not a code walker issue, but something to do with make.

**Comment by JanWielemaker** [comment]
I'm not aware of this.   Try `?- profile(make).` to see what it is doing or try to get more details from others on our forum.   Checking for undefined predicates has changed, but much longer ago than 6 months.   When did you upgrade between which versions?

**Comment by xpxaxsxi** [comment]
I start Swi-prolog threaded, 64 bits, version 8.3.24 without packs (no own files loaded).

I think some information is cached because succeeding profile(make)'s are faster.

Then I profile(prolog_walk_code([]). It walks all files, I think, and  this is possibly what it is intended.

Then I profile(prolog_walk_code([module(user]), and I get 0.000 seconds. 
I don't know the internals of Swi-Prolog, but I expect that code walker should walk only the user module??? 



```
1 ?- profile(make).
=====================================================================
Total time: 1.438 seconds
=====================================================================
Predicate                       Box Entries =    Calls+Redos     Time
=====================================================================
read_clause/3                         1,845 =    1,845+0        21.8%
access_file/2                           856 =      856+0         9.2%
exists_directory/1                      849 =      849+0         8.0%
$get_predicate_attribute/3           68,656 =   68,656+0         6.9%
$record_clause/3                      1,814 =    1,814+0         5.7%
$expand:is_meta_call/3                8,145 =    4,680+3,465     2.3%
arg/3                                27,097 =   15,488+11,609    2.3%
retractall/1                            446 =      446+0         2.3%
term_expansion/2                      2,464 =    2,080+384       2.3%
$autoload:do_autoload/3              87,139 =      138+87,001    1.1%
$expand:call_goal_expansion/5        15,532 =    7,626+7,906     1.1%
$expand:contains_functions2/1        19,301 =    9,097+10,204    1.1%
$expand:expand_control/8             58,499 =    1,633+56,866    1.1%
$expand:expand_functional_notation/7 28,406 =    4,550+23,856    1.1%
expand_term/4                        86,178 =    1,993+84,185    1.1%
$expand:expand_terms/5              157,886 =    4,739+153,147   1.1%
$expand:f2_pos/6                      4,659 =    4,490+169       1.1%
$expand:rename/3                      2,240 =    2,240+0         1.1%
$syspreds:define_or_generate/1       76,225 =   27,709+48,516    1.1%
predicate_property/2                 76,670 =   27,780+48,890    1.1%
$syspreds:property_predicate/2       76,670 =   27,780+48,890    1.1%
prolog_codewalk:find_walk_from_module/2132,408 =       31+132,377   1.1%
prolog_codewalk:walk_called/4       161,923 =    2,472+159,451   1.1%
prolog_codewalk:walk_called_by_pred/2129,927 =    1,971+127,956   1.1%
inferred_meta_predicate/2            12,311 =    6,117+6,194     1.1%
true.
```

```
2 ?- profile(make).
=====================================================================
Total time: 0.609 seconds
=====================================================================
Predicate                       Box Entries =    Calls+Redos     Time
=====================================================================
$syspreds:define_or_generate/1       39,660 =   32,002+7,658    13.9%
$get_predicate_attribute/3           74,363 =   74,363+0        13.9%
prolog_codewalk:walk_called/4        92,084 =    2,903+89,181   11.1%
skip/2                                4,078 =    4,078+0         8.3%
prolog_codewalk:walk_option_autoload/28,001 =    8,001+0         5.6%
access_file/2                           165 =      165+0         5.6%
read_clause/3                           112 =      112+0         5.6%
$expand:extended_pos/3                   95 =       95+0         2.8%
$syspreds:module_property/1           1,313 =    1,313+0         2.8%
predicate_property/2                 39,921 =   32,018+7,903     2.8%
prolog:called_by/2                    8,005 =    8,005+0         2.8%
prolog_codewalk:walk_called_by_body/382,525 =    2,614+79,911    2.8%
prolog_codewalk:walk_meta_call/7      8,848 =      659+8,189     2.8%
prolog_codewalk:walk_option_evaluate/28,592 =    8,592+0         2.8%
prolog_codewalk:walk_option_trace_reference/28,031 =    8,031+0         2.8%
$time_source_file/3                     144 =      114+30        2.8%
clause/3                              3,110 =    1,323+1,787     2.8%
exists_directory/1                      108 =      108+0         2.8%
open/3                                   25 =       25+0         2.8%
throw/1                               3,131 =      108+3,023     2.8%
$autoload/1                           7,323 =       16+7,307     0.0%
$autoload:$autoload2/1                7,307 =       16+7,291     0.0%
$autoload:$autoload3/1                7,307 =       16+7,291     0.0%
$define_predicate/1                   7,371 =       16+7,355     0.0%
$find_library/5                         237 =       80+157       0.0%
true.
```

```
2 ?- profile(make).
=====================================================================
Total time: 0.484 seconds
=====================================================================
Predicate                       Box Entries =    Calls+Redos     Time
=====================================================================
prolog_codewalk:walk_called/4        84,713 =    2,903+81,810   17.2%
$syspreds:define_or_generate/1       32,297 =   32,010+287      10.3%
clause/3                              3,106 =    1,319+1,787    10.3%
predicate_property/2                 32,558 =   32,026+532       6.9%
prolog_source:seek_to_line/2             14 =        7+7         6.9%
$get_predicate_attribute/3           74,102 =   74,102+0         6.9%
functor/3                             8,220 =    8,220+0         6.9%
$predicate_property/2                31,969 =   31,969+0         3.4%
prolog:called_by/2                    8,005 =    8,005+0         3.4%
prolog_codewalk:evaluate/3            8,031 =    8,031+0         3.4%
prolog_codewalk:find_walk_from_module/284,678 =       40+84,638    3.4%
inferred_meta_predicate/2            14,353 =    7,138+7,215     3.4%
$c_current_predicate/2                3,074 =      153+2,921     3.4%
>/2                                   4,149 =    4,149+0         3.4%
compound_name_arity/3                 5,846 =    5,846+0         3.4%
memberchk/2                             541 =      541+0         3.4%
$find_library/5                         237 =       80+157       0.0%
$update_library_index/0                 188 =        1+187       0.0%
$autoload:autoload_from/3                75 =       15+60        0.0%
$autoload:autoload_in/2                 125 =      125+0         0.0%
$autoload:autoload_in/3                 125 =      125+0         0.0%
$autoload:autoloadable/2                120 =       15+105       0.0%
$autoload:current_autoload/3             30 =       30+0         0.0%
$autoload:goal_name_arity/3              30 =       30+0         0.0%
$autoload:index_checked_at/1             77 =       77+0         0.0%
true
```

```
3 ?- profile(prolog_walk_code([])).
=====================================================================
Total time: 0.469 seconds
=====================================================================
Predicate                       Box Entries =    Calls+Redos     Time
=====================================================================
$get_predicate_attribute/3           74,265 =   74,265+0        44.8%
$predicate_property/2                33,357 =   33,357+0        10.3%
prolog_codewalk:walk_called/4        83,751 =    3,157+80,594   10.3%
$syspreds:define_or_generate/1       33,549 =   33,381+168       6.9%
predicate_property/2                 33,549 =   33,381+168       6.9%
clause/3                              3,106 =    1,319+1,787     6.9%
prolog_codewalk:evaluate/3            8,435 =    8,435+0         3.4%
prolog_codewalk:set_evaluate_of_walk_option/32,418 =    2,418+0         3.4%
prolog_codewalk:walk_called_by_multifile/24,353 =      102+4,251     3.4%
prolog_metainference:inferred_meta_pred/311,474 =   11,474+0         3.4%
$find_library/5                          72 =       24+48        0.0%
$autoload:autoload_in/2                  24 =       24+0         0.0%
$autoload:autoload_in/3                  24 =       24+0         0.0%
$autoload:index_checked_at/1             24 =       24+0         0.0%
$autoload:library_index/3                72 =       72+0         0.0%
$autoload:load_library_index/2           48 =       24+24        0.0%
$autoload:load_library_index/3           48 =       24+24        0.0%
$autoload:load_library_index_p/1         24 =       24+0         0.0%
$syspreds:$clause_property/2          1,404 =    1,404+0         0.0%
$syspreds:canonical_source_file/2       296 =        9+287       0.0%
clause_property/2                     1,404 =    1,404+0         0.0%
current_module/1                         41 =        1+40        0.0%
current_predicate/2                     168 =       24+144       0.0%
$syspreds:module_property/1           1,327 =    1,327+0         0.0%
module_property/2                     1,817 =    1,817+0         0.0%
true.
```
```
4 ?- profile(prolog_walk_code([module(user)])).
=====================================================================
Total time: 0.000 seconds
=====================================================================
Predicate                       Box Entries =    Calls+Redos     Time
=====================================================================
$syspreds:$clause_property/2             33 =       33+0         0.0%
$predicate_property/2                   108 =      108+0         0.0%
clause_property/2                        33 =       33+0         0.0%
current_module/1                          1 =        1+0         0.0%
$syspreds:define_or_generate/1          108 =      108+0         0.0%
$syspreds:module_property/1               9 =        9+0         0.0%
module_property/2                        42 =       42+0         0.0%
predicate_property/2                    108 =      108+0         0.0%
$syspreds:property_module/2               9 =        9+0         0.0%
$syspreds:property_predicate/2          108 =      108+0         0.0%
error:has_type/2                         14 =       14+0         0.0%
must_be/2                                14 =       14+0         0.0%
prolog:called_by/2                       20 =       20+0         0.0%
prolog:called_by/4                       20 =       20+0         0.0%
prolog_codewalk:clause_not_from_development/477 =       20+57        0.0%
prolog_codewalk:default_walk_option/1     1 =        1+0         0.0%
prolog_codewalk:evaluate/3               20 =       20+0         0.0%
prolog_codewalk:find_walk_from_module/2  72 =        1+71        0.0%
prolog_codewalk:hidden_predicate/2       26 =       26+0         0.0%
prolog_codewalk:infer_new_meta_predicates/23 =        1+2         0.0%
prolog_codewalk:init_goal_in_scope/3      9 =        1+8         0.0%
prolog_codewalk:is_meta/1                 1 =        1+0         0.0%
prolog_codewalk:make_walk_option/3        1 =        1+0         0.0%
prolog_codewalk:multifile_predicate/3    21 =       21+0         0.0%
prolog_codewalk:no_walk_property/1        1 =        1+0         0.0%
true.
```

**Comment by JanWielemaker** [comment]
Thanks.  Fixed with 2fcc985af6bc9f4f40d1233d666a13d3ff341fa4.   The issue was that restricting the scan to user modules did not work, scanning all system modules as well.  This causes a lot of additional code to be loaded and scanned.   I didn't notice because I have a pretty fast machine and a local declaration that enables scanning the system modules always as I'm editing them often :smile:
